CRM/Core/ManagedEntities.php

```diff
@@ -236,6 +236,14 @@ public function reconcileUnknownModules() {
    *   Entity specification (per hook_civicrm_managedEntities).
    */
   public function insertNewEntity($todo) {
+    // If you have an existing entity and you want to convert it to managed you have to check if it already exists.
+    // Otherwise it will crash with DB error duplicate entry (eg. for PaymentProcessorType)
+    $existingEntity = civicrm_api3($todo['entity'], 'get', $todo['params']);
```

Comment:
@mattwire This is pretty clever, and it's nice how it side-steps the question of "What fields need to be matched?". It looks strictly safe as-is, but I'm not sure it's fully fit for the purpose you set out.... or maybe it merits a few disclaimers.

* It will only work for entities with schematic-consistency for the get+create fields. (This is true for many fields/entities, so it's probably true for most cases where you'd want hook_managed...)
* It may sometimes be a bit particular about how one upgrades. (Though perhaps that could be addressed with [options.match](https://docs.civicrm.org/dev/en/latest/api/v3/options/#optionsmatch)? More in a moment)

To see the latter point, let me take the `redsys` example and propose a hypothetical timeline of releases:

* v1.0 does an explicit `PaymentProcessorType.create`. Among the various params, it sets `'class_name' => 'Payment_Redsys',`
* v1.1 switches to `hook_managed` (to take advantage of implicit `create`/`update/`delete`), but it doesn't make any substantive change. All the params match-up.
* v2.0 still uses `hook_managed`, but it changes one param -- `'class_name' => 'Payment_Redsys2'`.

If the upgrade path is v1.0=>v1.1=>v2.0, then it should find `$existingEntity` as expected. But if you upgrade v1.0=>v2.0, then it needs to find `$existingEntity`, but it won't because some params (`class_name`) don't match. The problem is that `PaymentProcessorType, ger, $todo['params']` matches on every field. You really just want to match on `name`.

For APIv3, one can set a matching-rule like `$params['options']['match'] = 'name'`. So if you set that in the `hook_managed` params, then it just might work. Well... ok, the `get` operation (per se) doesn't understand `options.match`... but perhaps it'd work if either:

* Skip this patch. In your managed-entity (v1.1, v2.0, etc), specify `options.match=name`. In theory, when it goes to create the entity, it should match the existing one by name... then update and take ownership of it.
* Proceed with this patch, but make it conditional on `options.match`, eg
    ```php
    $existingEntity = isset($todo['params']['options']['match'])
      ? civicrm_api3($todo['entity'], 'get', CRM_Utils_Array::subset($todo['params'], $todo['params']['options']['match']))
      : NULL;
    ```

I think you'll find the various migration/upgrade paths will be more robust if you can get it to match more narrowly by `name`.